emData/generate_MC.py

```diff
@@ -0,0 +1,170 @@
+        FM = re.sub(r"FM_(....)_......", r"\1", memName)
+        if tcName not in FMMems:
+            FMMems[tcName] = []
+        FMMems[tcName].append(FM)
+wiresFile.close()
+
+# Open and print out preambles for the parameters and top files.
+parametersFile = open("MatchCalculator_parameters.h", "w")
+parametersFile.write(
+    "#ifndef TrackAlgorithm_MatchCalculator_parameters_h\n"
+    "#define TrackAlgorithm_MatchCalculator_parameters_h\n"
+    "\n"
+    "// This file contains numbers of memories and bit masks that are specific to\n"
+    "// each MatchCalculator and that come directly from the wiring.\n"
+    "//\n"
+    "// The validity of each of the barrel CM memories is determined by\n"
+    "// FMMask. The bits of this mask, from least significant to most\n"
+    "// significant, represent the memories in the order they are passed to\n"
+    "// MatchCalculator; e.g., the LSB corresponds to\n"
+    "// TF::L1L2. If a bit is set, the corresponding memory is\n"
+    "// valid, if it is not, the corresponding memory is not valid.\n"
+)
+topHeaderFile = open("MatchCalculatorTop.h", "w")
+topHeaderFile.write(
+    "#ifndef TrackAlgorithm_MatchCalculatorTop_h\n"
+    "#define TrackAlgorithm_MatchCalculatorTop_h\n"
+    "\n"
+    "#include \"MatchCalculator.h\"\n"
+)
+topFile = open("MatchCalculatorTop.cc", "w")
+topFile.write(
+    "#include \"MatchCalculatorTop.h\"\n"
+    "\n"
+    "////////////////////////////////////////////////////////////////////////////////\n"
+    "// Top functions for various MatchCalculators (MC). For each iteration of\n"
+    "// the main processing loop, a MC retrieves an array of Candidate Matches,\n"
+    "// and computes Full Matches from each.\n"
+    "////////////////////////////////////////////////////////////////////////////////\n"
+)
+
+# Calculate parameters and print out parameters and top function for each MC.
+for tcName in sorted(CMMems.keys()):
+    seed = re.sub(r"MC_(..)....", r"\1", tcName)
+    iMC = re.sub(r"MC_.....(.)", r"\1", tcName)
+
+    # numbers of memories
+    nCMMem = len(CMMems[tcName])
+    FMMask = 0
+    for FM in FMMems[tcName]:
+        FMMask = FMMask | (1 << TF_index[FM])
+
+    # Print out parameters for this MC.
+    parametersFile.write(
+        "\n"
+        "// magic numbers for " + tcName + "\n"
+        "template<> constexpr uint32_t FMMask<TF::" + seed + ", MC::" + iMC + ">() {\n"
+        "  return 0x%X;\n"
+        "}\n" % FMMask
+    )
+    parametersFile.write(
+        "template<TF::layerDisk Layer, MC::imc PHI, TF::seed Seed> constexpr bool FMMask() {\n"
+        "  return FMMask<Layer, PHI>() & (1<<Seed);\n"
+        "}\n"
+    )
+
+    # Print out prototype for top function for this MC.
+    topHeaderFile.write(
+        "\n"
+        "constexpr int maxMatchCopies(" + str(len(TF_index)) + ");\n"
+        "constexpr int maxFullMatchCopies(" + str(len(TF_index)) + ");\n"
+        "\n"
+        "void MatchCalculator_" + seed + iMC + "(\n"
```

Comment:
@bryates This is nitpicky, but could you use the same suffix style as is used in the wiring, e.g. something like `MatchCalculator_L3PHIC` instead of `MatchCalculator_L3C`?

Reply:
Good catch!